src/Raven.Client/Documents/Session/DocumentSession.cs

```diff
@@ -137,15 +139,48 @@ public bool Exists(string id)
         /// <param name="entity">The entity.</param>
         public void Refresh<T>(T entity)
         {
-            DocumentInfo documentInfo;
-            if (DocumentsByEntity.TryGetValue(entity, out documentInfo) == false)
+            if (DocumentsByEntity.TryGetValue(entity, out var documentInfo) == false)
                 throw new InvalidOperationException("Cannot refresh a transient instance");
             IncrementRequestCount();
 
             var command = new GetDocumentsCommand(new[] { documentInfo.Id }, includes: null, metadataOnly: false);
             RequestExecutor.Execute(command, Context, sessionInfo: _sessionInfo);
 
-            RefreshInternal(entity, command, documentInfo);
+            var commandResult = (BlittableJsonReaderObject)command.Result.Results[0];
+            RefreshInternal(entity, commandResult, documentInfo);
+        }
+
+        /// <summary>
+        /// Refreshes the specified entities from Raven server.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="entities">The entities.</param>
+        public void Refresh<T>(IEnumerable<T> entities)
+        {
+            List<string> ids = new();
+            List<DocumentInfo> documentInfos = new();
+
+            foreach (var entity in entities)
+            {
+                if (DocumentsByEntity.TryGetValue(entity, out var docInfo) == false)
+                    throw new InvalidOperationException("Cannot refresh a transient instance");
+                documentInfos.Add(docInfo);
+                ids.Add(docInfo.Id);
+            }
+
+            IncrementRequestCount();
+
+            var command = new GetDocumentsCommand(ids.ToArray(), includes: null, metadataOnly: false);
+            RequestExecutor.Execute(command, Context, sessionInfo: _sessionInfo);
+
+            var j = 0;
+            foreach (var entity in entities)
+            {
+                var commandResult = (BlittableJsonReaderObject)command.Result.Results[j];
```

Comment:
what will happen here, if we refresh list with same entities (like `await session.Advanced.RefreshAsync(new[] { sd, sd, sd }.AsEnumerable());`)

Reply:
checked it, and it will break. I applied a fix to handle this situation.
therefore,
Done